lib/src/Uncertainty/Bayesian/UserDefinedMetropolisHastings.cxx

```diff
@@ -0,0 +1,157 @@
+//                                               -*- C++ -*-
+/**
+ *  @brief Metropolis-Hastings algorithm
+ *
+ *  Copyright 2005-2022 Airbus-EDF-IMACS-ONERA-Phimeca
+ *
+ *  This library is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Lesser General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "openturns/UserDefinedMetropolisHastings.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+
+BEGIN_NAMESPACE_OPENTURNS
+
+
+CLASSNAMEINIT(UserDefinedMetropolisHastings)
+
+static const Factory<UserDefinedMetropolisHastings> Factory_UserDefinedMetropolisHastings;
+
+/* Default constructor */
+UserDefinedMetropolisHastings::UserDefinedMetropolisHastings()
+  : MetropolisHastingsImplementation()
+{
+  // Nothing to do
+}
+
+
+/* Parameters constructor */
+UserDefinedMetropolisHastings::UserDefinedMetropolisHastings(const Distribution & targetDistribution,
+    const Point & initialState,
+    const Distribution & proposal,
+    const Function & linkFunction,
+    const Indices & marginalIndices)
+  : MetropolisHastingsImplementation(targetDistribution, initialState, marginalIndices)
+{
+  setProposalAndLinkFunction(proposal, linkFunction);
+}
+
+/* Parameters constructor */
+UserDefinedMetropolisHastings::UserDefinedMetropolisHastings(const Function & targetLogPDF,
+    const Domain & support,
+    const Point & initialState,
+    const Distribution & proposal,
+    const Function & linkFunction,
+    const Indices & marginalIndices)
+  : MetropolisHastingsImplementation(targetLogPDF, support, initialState, marginalIndices)
+{
+  setProposalAndLinkFunction(proposal, linkFunction);
+}
+
+
+/* String converter */
+String UserDefinedMetropolisHastings::__repr__() const
+{
+  return OSS() << "class=" << UserDefinedMetropolisHastings::GetClassName()
+         << " name=" << getName()
+         << " derived from " << MetropolisHastingsImplementation::__repr__()
+         << " proposal=" << proposal_;
+}
+
+
+UserDefinedMetropolisHastings* UserDefinedMetropolisHastings::clone() const
+{
+  return new UserDefinedMetropolisHastings(*this);
+}
+
+
+Point UserDefinedMetropolisHastings::getCandidate() const
+{
+  // proposal_ distribution conditioned on the currentState_
+  proposal_.setParameter(proposalLinkFunction_(currentState_));
```

Comment:
This line is the sensitive point of the code: for the user, knowing that this is what the linkFunction is used for is important.

Reply:
This is now explained at the start of the API doc page: see my comment https://github.com/openturns/openturns/pull/2035#discussion_r941561754